2FA Removal Feature  🏷

dlipeles opened the pull request:
### 2FA Removal Feature 🔐 
![Group](https://user-images.githubusercontent.com/4713771/55503150-1f8e4000-561c-11e9-9f59-14a27d6570e6.png)

#### Features 
This pull request fulfills the `2fa Removal Spec` for publishers who have lost their second factor authenticator, but wish to regain access to their account. 

* Options - Provides an alternative to creating a new account and transferring channels.
* Secure Flow - Enforces a waiting period & requires re-verification of channels/wallet.
* Self Service - Publishers can initiate the removal at their convenience.

#### Notes
* Admins can see the progress of a 2fa removal from the publisher page.
* There is a `2 week waiting period` before 2fa removal , then a `4 week lockout` on payments for the publisher's security.
* Admins can preempt a 2fa removal or lockout from the admin dashboard.

#### How To Test

#### Setup
* Set the `TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD` to `300` seconds and `LOCKED_STATUS_WAITING_PERIOD` to `300` seconds. 
* Set the cron job in sidekiq to every minute `* * * * *`.
* Create a publisher with `2fa enabled`, `channels`, and a `wallet`.
* Request the 2fa removal from the landing page, confirm the e-mail. 

#### In Action
* Log in as an `Admin` to see progress on the job. 
* You should see the following sequence: 
1. `2fa job registered` 
2. `Reminder e-mails sent`
3. `Publisher 2fa removed`, `Publisher channels removed`, `Publisher wallet disconnected`, `Publisher state becomes LOCKED`
4. `2fa job completed`, `Publisher state becomes ACTIVE`

corymcdonald commented:
I wonder if it should be more streamlined? Like after the user logs in and they ca't get passed the 2 FA screen we already have their email so I don't know if the user necessarily needs to enter their email?

yachtcaptain23 commented on app/controllers/publishers_controller.rb:
```diff
@@ -199,6 +203,17 @@ def update
+      flash[:notice] = t("publishers.two_factor_authentication_removal.request_success")
+      MailerServices::TwoFactorAuthenticationRemovalRequestEmailer.new(publisher: publisher).perform
+    else
+      flash[:warning] = t("publishers.two_factor_authentication_removal.request_not_found")
```
This exposes the email addresses that we have. We should change the message to the same as the above but only run `MailerServices::TwoFactorAuthenticationRemovalRequestEmailer.new(publisher: publisher).perform` if the publisher exists

dlipeles replied:
Thanks for pointing this out @yachtcaptain23 , I made the fix.

yachtcaptain23 commented on test/fixtures/two_factor_authentication_removals.yml:
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Unused fixture

yachtcaptain23 commented on test/models/two_factor_authentication_removal_test.rb:
```diff
@@ -0,0 +1,9 @@
+require "test_helper"
+
+describe TwoFactorAuthenticationRemoval do
```
Would you add tests to make sure the before and after state are correct?

yachtcaptain23 commented on app/jobs/two_factor_authentication_removal_job.rb:
```diff
@@ -0,0 +1,33 @@
+        # Delete 2fa registrations, channels, and disconnect wallet
+        publisher.totp_registration.destroy if publisher.totp_registration.present?
+        publisher.u2f_registrations.destroy_all if !publisher.u2f_registrations.empty?
+        publisher.channels.destroy_all if !publisher.channels.empty?
```
Users need to still be paid. Please don't destroy, but you can remove the `publisher_id` on the `channels` table.

dlipeles replied:
Agreed, the users still need to, and will be paid; @evq's insight here was really helpful to me https://bravesoftware.slack.com/archives/C8GJLTQHW/p1552073068041900

The browser will hold funds until the publisher re-verifies their channels, for that reason we don't want channels to appear in the publishers list until they re-verify.

yachtcaptain23 commented on app/models/two_factor_authentication_removal.rb:
```diff
@@ -0,0 +1,49 @@
+    end
+  end
+
+  def two_factor_authentication_removal_time_completed?
```
Prefer to bring this down to one line: 
`(Time.now - created_at) >= TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD`

yachtcaptain23 commented on app/models/two_factor_authentication_removal.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  def locked_status_time_completed?
+    time = Time.now
```
Same here

yachtcaptain23 commented on app/models/two_factor_authentication_removal.rb:
```diff
@@ -0,0 +1,45 @@
+
+  # Returns time remaining in seconds
+  def total_time_remaining
+    time = Time.now
```
Prefer to be more ruby-like to simply be like `TOTAL_WAITING_PERIOD - (Time.now - created_at)`

yachtcaptain23 commented on app/services/mailer_services/two_factor_authentication_removal_request_emailer.rb:
```diff
@@ -0,0 +1,23 @@
+module MailerServices
+    class TwoFactorAuthenticationRemovalRequestEmailer < BaseService
```
Would you fix the indentation to be 2 spaces?

yachtcaptain23 commented on app/services/payout_report_publisher_includer.rb:
```diff
@@ -6,7 +6,8 @@ def initialize(payout_report:, publisher:, should_send_notifications:)
   def perform
-    return if !@publisher.has_verified_channel? || @publisher.excluded_from_payout?
+    return if !@publisher.has_verified_channel? || @publisher.locked? || @publisher.excluded_from_payout?
+    publisher_has_unsettled_balance = false
```
Extra non-intended line?

dlipeles replied:
Fixed, will keep some spacing for readability

yachtcaptain23 commented on app/jobs/two_factor_authentication_removal_job.rb:
```diff
@@ -0,0 +1,33 @@
+      # 14th day of the job
+      elsif two_factor_authentication_removal.two_factor_authentication_removal_time_completed? && !two_factor_authentication_removal.removal_completed?
+        # Delete 2fa registrations, channels, and disconnect wallet
+        publisher.totp_registration.destroy if publisher.totp_registration.present?
```
Would you lock this block into one transaction?

yachtcaptain23 commented on app/jobs/two_factor_authentication_removal_job.rb:
```diff
@@ -0,0 +1,33 @@
+
+      # After 6 weeks of the job. Restore publisher to Active state, delete the two_factor_authentication_removal object.
+      elsif two_factor_authentication_removal.locked_status_time_completed?
+        publisher.status_updates.create(status: PublisherStatusUpdate::ACTIVE)
```
One transaction here too

dlipeles commented:
Added the requested feedback, I also added a new `Security` tab on the admin dash for 2fa related information 

![Screen Shot 2019-04-05 at 4 00 10 PM](https://user-images.githubusercontent.com/4713771/55653299-1045f880-57bc-11e9-983f-ac4ee7315627.png)

yachtcaptain23 commented on app/jobs/two_factor_authentication_removal_job.rb:
```diff
@@ -0,0 +1,36 @@
+        ActiveRecord::Base.transaction do
+          publisher.totp_registration.destroy if publisher.totp_registration.present?
+          publisher.u2f_registrations.destroy_all if !publisher.u2f_registrations.empty?
+          publisher.channels.destroy_all if !publisher.channels.empty?
```
Actually would you run `DeletePublisherChannelJob`? This makes a call to promo server to remove the unused channel.

corymcdonald commented on app/controllers/publishers_controller.rb:
```diff
@@ -17,7 +17,12 @@ class PublishersController < ApplicationController
-  before_action :authenticate_publisher!, except: %i(show)
+  before_action :authenticate_publisher!, except: %i(
+    show
+    two_factor_authentication_removal
```
In my mind this is kind of clear indication that these methods likely belongs in another controller. This controller relates to a publisher and almost all of these methods require authentication via publisher. 

The reasoning I have for moving it into a new controller is
* they don't require any publisher to be logged in
* relates to a 2FA "model" as in everything about these can be mapped to create, and update

corymcdonald commented on app/controllers/publishers_controller.rb:
```diff
@@ -109,6 +114,35 @@ def update
   end
 
+  def request_two_factor_authentication_removal
+    publisher = Publisher.find_by_email(params[:email])
```
Can you check to see if the case is sensitive on this?

corymcdonald commented on app/controllers/publishers_controller.rb:
```diff
@@ -109,6 +114,35 @@ def update
+      flash[:notice] = t("publishers.two_factor_authentication_removal.request_success")
+      MailerServices::TwoFactorAuthenticationRemovalRequestEmailer.new(publisher: publisher).perform
+    else
+      flash[:notice] = t("publishers.two_factor_authentication_removal.request_success")
```
If the publisher doesn't exist we will say the same thing to the user saying that the request was success?

corymcdonald commented on app/models/two_factor_authentication_removal.rb:
```diff
@@ -0,0 +1,38 @@
+  include ActionView::Helpers::DateHelper
+  belongs_to :publisher
+  # 2 Weeks represented in seconds
+  TWO_FACTOR_AUTHENTICATION_REMOVAL_WAITING_PERIOD = 1209600
```
Let's just do some ✨ Rails Magic 🌈 

`2.weeks.seconds`
`4.weeks.seconds`
`6.weeks.seconds`

corymcdonald commented on app/models/two_factor_authentication_removal.rb:
```diff
@@ -0,0 +1,38 @@
+    if total_seconds <= LOCKED_STATUS_WAITING_PERIOD
+      distance_of_time_in_words(total_seconds)
+    else
+      "Not started yet"
```
Internationalization

corymcdonald commented on app/services/payout_report_publisher_includer.rb:
```diff
@@ -6,10 +6,10 @@ def initialize(payout_report:, publisher:, should_send_notifications:)
-    return if !@publisher.has_verified_channel? || @publisher.excluded_from_payout?
+    return if !@publisher.has_verified_channel? || @publisher.locked? || @publisher.excluded_from_payout?
 
+    publisher_has_unsettled_balance = false
```
we assigning this twice?

corymcdonald commented on config/routes.rb:
```diff
@@ -158,11 +161,13 @@
         patch :approve_channel
         get :statement
         post :create_note
+        get :cancel_two_factor_authentication_removal
```
Really think these should just be routes on a `resources :two_factor_authentication_removal` controller

dlipeles replied:
You are right, I will revisit this in a later PR and make this refactor
